mlir/lib/Dialect/MIOpen/Transforms/ThreadwiseGemmLowering.cpp

```diff
@@ -274,31 +272,27 @@ struct XdlopsGemmV2RewritePattern : public OpConversionPattern<XdlopsGemmV2Op> {
     Type bufferAElementType = bufferAType.getElementType();
     Type bufferBElementType = bufferBType.getElementType();
 
+    int64_t KPerThread = IsKReduction ? K / num_input_blks : K;
+
     if (!IsKReduction) {
-      // store bufferA logic.
 
-      // Original C++ logic.
-      // static_if<!IsKReduction>{}([&](auto) {
-      //   for(index_t m_i = 0; m_i < MRepeats; ++m_i)
-      //     for(index_t k_i      = 0; k_i < K; ++k_i)
-      //       a[k_i + m_i * K] = p_a_wave[k_i * M + laneId + MPerXdlops * m_i];
-      // p_a_wave need to be offseted by waveOffsetA.
+      // store bufferA logic.
+      // for(index_t m_i = 0; m_i < MRepeats; ++m_i)
+      //   for(index_t k_i      = 0; k_i < K; ++k_i)
+      //     a[k_i + m_i * K] = p_a_wave[k_i * M + laneId + MPerXdlops * m_i];
+      // Note: p_a_wave need to be offseted by waveOffsetA.
 
       auto outerLoopM = b.create<AffineForOp>(loc, 0, MRepeats);
-      auto olmb = ConversionPatternRewriter::atBlockTerminator(
-          outerLoopM.getBody(), b.getListener());
+      auto olmb = OpBuilder::atBlockBegin(outerLoopM.getBody());
```

Comment:
First Just trying to clarify here: Why does debug not work using `OpBuilder`? Does it not populate debug messages in the `LLVM_DEBUG`? Wants to understand the symptom to it so I can identify this better.

If this indeed break debug in certain way, I'd still prefer `OpBuilder::atBlockBegin(outerLoopM.getBody(), b.getListener());`, because `atBlockBegin` is a child class function that eventually fall back on `OpBuilder`. Since it is a static function anyway, why not use the parent `OpBuilder` version of this?

Reply:
Ok, so the situation is that pattern rewriting itself has debug messages - these are handled through the listener methods on the rewriter. Creating new builders/rewriters without preserving the listener field breaks debug.

Second, I want to use the child class's version both in case it does something fancy, so we don't erase the actual type of the child, and so we have access to the conversion-rewriter methods in their context.

I agree with `atBlockBegin`, but I want that to be `ConversionPatternRewriter::atBlockBegin(..., b.getListener())`

And lastly, eventually, we should move to insertion guards and just using one builder object, but that might not be a refactor for this PR. (Last I checked, builder construction is not guaranteed to be cheap and so setting the insertion point on an existing builder is prefered)

Reply:
> these are handled through the listener methods on the rewriter

Good to know. I didn't notice this before

> I agree with atBlockBegin

Sounds good.

> Second, I want to use the child class's version both in case it does something fancy

I was not accurate in previous comment. Actually, this `ConversionPatternRewriter` does not reroute a implementation of `atBlockBegin()` at all, and it inherits from the `OpBuilder` interface. Searching through the entire code base I don't find another instance where some consumer would ask for `atBlockBegin()` from a rewriter, and all instances of its use are from the `OpBuilder`.

At this point I don't think this matters to invoke either - it is likely by design insertion point setting and getting are from `OpBuilder` (and rewriter will keep to be orthalganal) . But I'd prefer `OpBuilder`'s version to be consistent with the rest of the code base.

Reply:
The rest of the code base got rather systematically adjusted to use `ConversionPatternRewriter::` in places where there's indeed a `ConversionPatternRewriter`, and I'd like to not have the static type of the object we're using for building get changed out from under us. I strongly prefer that, if we start with a `ConversionPatternRewriter` as the argument to our `matchAndRewrite`, we stick with a `ConversionPatternRewriter` throughout (except in utility methods that can downgrade, but then C++ is doing the right thing and adding vtables and so on).

Reply:
That is, when I was turning this into conversion patterns, I had some very weird bugs arise from using a `ConversionPatternRewriter` to create an `OpBuilder`, and so I want to keep that consistency.

Reply:
No problem, thanks for doing this massive cleanup!